chips/nrf52/src/comp.rs

```diff
@@ -0,0 +1,301 @@
+            Internal1V8 = 1,
+            // VREF = 2.4V (VDD >= VREF + .2V)
+            Internal2V4 = 2,
+            // VREF = VDD
+            VDD = 4,
+            // Select another pin as a reference
+            // (VDD >= VREF >= AREFMIN)
+            AnalogReference = 5
+        ]
+    ],
+    // If in diff mode, or single-ended mode with an analog reference,
+    // use the pin specified here for VIN-
+    ExternalRefSelect[
+        ExternalRefSelect OFFSET(0) NUMBITS(3) [
+            AnalogRef0 = 0,
+            AnalogRef1 = 1,
+            AnalogRef2 = 2,
+            AnalogRef3 = 3,
+            AnalogRef4 = 4,
+            AnalogRef5 = 5,
+            AnalogRef6 = 6,
+            AnalogRef7 = 7
+        ]
+    ],
+    Mode[
+        // Controls power usage and correspondingly speed of comparator
+        SpeedAndPower OFFSET(0) NUMBITS(3) [
+            Low = 0,
+            Normal = 1,
+            High = 2
+        ], 
+        OperatingMode OFFSET(8) NUMBITS(1) [
+            SingleEnded = 0,
+            Differential = 1
+        ]
+    ],
+    // HYST register for hysteresis in diff mode
+    // Turns on 50mV hysteresis
+    Hysteresis [
+        Hysteresis OFFSET(0) NUMBITS(1)[]
+    ]
+];
+
+pub struct Comp {
+    registers: StaticRef<CompRegisters>,
+    client: OptionalCell<&'static dyn analog_comparator::Client>,
+}
+
+impl Comp {
+    const fn new(registers: StaticRef<CompRegisters>) -> Comp {
+        Comp {
+            registers: registers,
+            // state: Cell::new(State::Idle),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn set_client(&self, client: &'static dyn analog_comparator::Client) {
+        self.client.set(client);
+    }
+
+    // Returns if comparator was already enabled
+    fn enable(&self) {
+        let regs = &*self.registers;
+        // Checks if it's already enabled
+        // Assumes no one else is writing to comp regs directly
+        if regs.enable.matches_any(Enable::ENABLE::Enabled) {
+            return;
+        }
+
+        // Set mode to Differential
+        // Differential and single ended are pretty much the same,
+        // except single-ended gives more options for input and
+        // uses a ref ladder for hysteresis instead of a set voltage
+        regs.mode
+            .write(Mode::OperatingMode::Differential + Mode::SpeedAndPower::Normal);
+        // VIN+ = Pin 0
+        regs.psel.write(PinSelect::PinSelect::AnalogInput0);
+        // VIN- = Pin 1
+        regs.extrefsel
+            .write(ExternalRefSelect::ExternalRefSelect::AnalogRef1);
+        // Disable hysteresis
+        regs.hyst.write(Hysteresis::Hysteresis::CLEAR);
+
+        regs.enable.write(Enable::ENABLE::Enabled);
+        // start comparator
+        regs.events_ready.set(0);
+        regs.tasks_start.set(1);
+        // wait for comparator to be ready
+        // delay is on order of 3 microseconds so spin wait is OK
+        while regs.events_ready.get() == 0 {}
```

Comment:
Is there any situation where this will fail?

(See #1552 for more thoughts)

Reply:
Under normal working conditions, no. If the comparator is somehow broken then it might not be possible to enable it, in which case it could be a problem.